src/storage.rs

```diff
@@ -1,433 +1,480 @@
+/// performance for the components mostly present in entities.
+pub struct VecStorage<T>(Vec<T>);
 
+impl<T> UnprotectedStorage<T> for VecStorage<T> {
     fn new() -> Self {
-        VecStorage {
-            set: BitSet::new(),
-            values: InnerVec(Vec::new()),
-        }
+        VecStorage(Vec::new())
     }
-    fn get(&self, e: Entity) -> Option<&T> {
-        let id = e.get_id();
-        if self.set.contains(id as u32) {
-            let v = unsafe { self.values.0.get_unchecked(id) };
-            if v.generation == e.get_gen() {
-                return Some(&v.data);
+    unsafe fn clean<F>(&mut self, has: F) where F: Fn(Index) -> bool {
+        use std::mem;
+        for (i, v) in self.0.drain(..).enumerate() {
+            if !has(i as Index) {
+                // if v was not in the set the data is invalid
+                // and we must forget it instead of dropping it
+                mem::forget(v);
             }
         }
-        None
-
     }
-    fn get_mut(&mut self, e: Entity) -> Option<&mut T> {
-        let id = e.get_id();
-        if self.set.contains(id as u32) {
-            let v = unsafe { self.values.0.get_unchecked_mut(id) };
-            if v.generation == e.get_gen() {
-                return Some(&mut v.data);
-            }
-        }
-        None
+    unsafe fn get(&self, id: Index) -> &T {
+        self.0.get_unchecked(id as usize)
     }
-    fn insert(&mut self, e: Entity, mut v: T) {
-        use std::{ptr, mem};
-
-        let id = e.get_id();
-        if self.set.contains(id as u32) {
-            let mut data = &mut self.values.0[id];
-            data.generation = e.get_gen();
-            mem::swap(&mut data.data, &mut v);
-            Some(v)
-        } else {
-            self.set.add(id as u32);
-            if self.values.0.len() <= id {
-                self.extend(id);
-            }
-            unsafe {
-                ptr::write(
-                    &mut self.values.0[id],
-                    GenerationData{
-                        generation: e.get_gen(),
-                        data: v
-                    }
-                );
-            }
-            None
-        };
+    unsafe fn get_mut(&mut self, id: Index) -> &mut T {
+        self.0.get_unchecked_mut(id as usize)
     }
-    fn remove(&mut self, e: Entity) -> Option<T> {
-        use std::ptr;
-
-        let (id, gen) = (e.get_id(), e.get_gen());
-        let gen_matches = self.values.0.get(id)
-            .map(|x| x.generation == gen).unwrap_or(false);
-
-        if gen_matches && self.set.remove(id as u32) {
-            let value = unsafe { ptr::read(&self.values.0[id]) };
-            Some(value.data)
-        } else {
-            None
+    unsafe fn insert(&mut self, id: Index, v: T) {
+        let id = id as usize;
+        if self.0.len() <= id {
+            let delta = id + 1 - self.0.len();
+            self.0.reserve(delta);
+            self.0.set_len(id + 1);
         }
+        self.0[id] = v;
```

Comment:
You need to be careful here. if `T` supports drop this will drop the element before overwriting it, so it would drop an uninitialized value. Which is why I was using `ptr::write` before. of course if you use `ptr::write` and there was a value was suppose to be dropped you have now added a memory leak. If the value was set before you need to either use `mem::swap` to exchange it, or `ptr::read` and `ptr::write` to exchange it.